### PR TITLE
Improve validation and error handling in resolve_bet

### PR DESCRIPTION
**Pull request by linnbenton**
## Summary
This PR improves validation and error handling in the `resolve_bet` function.

## Problem
Previously, the contract accessed bets directly using:
self.bets[gl.message.sender_address][bet_id]

This could result in unclear errors when:
- the bet does not exist
- the bet does not belong to the caller

## Solution
Added explicit checks to ensure:
- the bet exists
- the bet belongs to the caller

## Changes
- Added validation for bet ownership and existence
- Improved error message clarity
- Minor code structure improvements

## Impact
No breaking changes. Improves reliability and developer experience.

## Summary by CodeRabbit

* **New Features**
  * Added a public bet placement action to create per-sender bets.

* **Bug Fixes**
  * Stronger validation of match results and bet ownership to prevent invalid resolutions.
  * Fixed initialization to avoid missing bet/score errors and ensured consistent point updates.

* **Refactor**
  * Standardized sender handling and normalized internal bet access.
  * Match-result parsing now requires valid JSON and enforces expected fields/types.

* **Tests**
  * Converted tests to pytest, added unit and integration resolution tests plus a placeholder test.

* **Chores**
  * Added local fallback for missing runtime bindings to improve testability.

### Before
resolve_bet could proceed with invalid or missing data.

### After
Validation ensures correct inputs before execution, preventing inconsistent state.

### CODE REVIEW

**Comment by linnbenton** [comment]
This change ensures safer access to bets by preventing implicit KeyError-like failures and providing clearer error messages when a bet is missing or not owned by the caller.

**Comment by linnbenton** [comment]
Also improved code formatting and readability for better maintainability.

**Comment by linnbenton** [comment]
Happy to make further improvements if needed. Thanks for reviewing!

**Comment by linnbenton** [comment]
Hi, just checking if any feedback or changes needed 🙏

**Comment by linnbenton** [comment]
This PR improves contract safety by adding stricter validation and preventing invalid states in resolve_bet.

✅ Prevents edge-case failures  
✅ Improves reliability for real users  
✅ Keeps logic backward-compatible  

Happy to extend with tests or additional validation if needed.

**Comment by linnbenton** [comment]
Added strict validation for match result and improved error handling.

Also added deterministic unit tests with mocking to avoid dependency on a running node.

All tests pass locally (6/6).

**Comment by linnbenton** [comment]
Hi maintainers 👋  
Just checking if there’s any feedback needed.  
Happy to update anything if required.

File: test/test_footbal_bet.py [hunk 1]
```diff
@@ -1,140 +1,117 @@
...
+@pytest.fixture
+def contract():
+    return FootballBets()
 
 
-def test_football_bets_success_win():
-    # Contract Deploy
+@pytest.fixture
+def user():
+    return Address("0x123")
```
[the remark this answers is not in the document]
Comment (linnbenton) [reply test/test_footbal_bet.py:17]: Good catch 👍

Added an integration-style test that uses the real contract flow 
(place_bet → resolve_bet) instead of directly injecting storage.

This improves coverage while keeping tests deterministic 
in the current environment.

**Comment by linnbenton** [comment]
### ✅ Test Results

All tests passing locally:

Screenshot_127

- 7 tests passed
- Includes integration-style test
- Changes committed and pushed

**Comment by linnbenton** [comment]
test: add integration-style test and improve contract validation
Screenshot_129

File: test/test_footbal_bet.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (linnbenton) [reply test/test_footbal_bet.py:17]: Thanks for the detailed guidance — this makes sense.

I attempted to run the gltest-based setup locally, but it looks like the full
gltest runtime/CLI isn't available in my current environment (no gltest command
or fixture exposed, despite genlayer-test being installed).

To improve coverage in the meantime, I added an integration-style test that
uses the real contract flow (place_bet → resolve_bet) instead of directly
injecting storage.

Happy to update this to a full gltest-based test once I can run the runtime
locally, or if there's a recommended setup for it.

[the remark this answers is not in the document]
Comment (linnbenton) [reply test/test_footbal_bet.py:17]: Thanks for the clarification — that helps a lot.

That makes sense regarding gltest requiring the Studio backend and not being purely a Python-side tool.

For now, I'll keep the current unit tests and the integration-style test
(place_bet → resolve_bet) as an interim solution to improve coverage
without direct storage injection.

Yes, it would be great if you could open a follow-up issue to track adding
a proper gltest-based integration test once the environment is set up.

Appreciate the guidance 🙌

[the remark this answers is not in the document]
Comment (linnbenton) [reply test/test_footbal_bet.py:17]: Thanks for creating the follow-up issue — really helpful 🙌

I’ll take a look and self-assign it. Would be great to extend this with a proper
gltest-based integration test once the local Studio setup is ready.

Appreciate all the guidance on this!

[the remark this answers is not in the document]
Comment (linnbenton) [reply test/test_footbal_bet.py:17]: Thanks again for all the help and guidance — really appreciate it 🙌

I’ll follow up on issue #42 once I have the Studio environment set up.

Looking forward to contributing more!

[the remark this answers is not in the document]
Comment (linnbenton) [reply test/test_footbal_bet.py:17]: Thanks a lot — really enjoyed working on this as well! 🙌

Glad the improvements to resolve_bet were helpful.
I’ll follow up on issue #42 once I get the Studio environment running.

Looking forward to contributing more 🚀

[the remark this answers is not in the document]
Comment (linnbenton) [reply test/test_footbal_bet.py:17]: Really appreciate the support and feedback throughout — it was a great experience 🙌

Glad the resolve_bet improvements landed well.
I’ll circle back to issue #42 once I have the Studio environment set up.

Looking forward to contributing more 🚀

[the remark this answers is not in the document]
Comment (linnbenton) [reply test/test_footbal_bet.py:17]: Really appreciate the kind words and all the guidance throughout — it made the whole process very smooth 🙌

Glad the resolve_bet improvements were useful.
I’ll revisit issue #42 once I have the Studio environment set up.

Looking forward to contributing more 🚀

[the remark this answers is not in the document]
Comment (linnbenton) [reply test/test_footbal_bet.py:17]: Thanks again — really appreciate the support throughout! 🙌

Looking forward to the next one 🚀

File: contracts/football_bets.py [hunk 2]
```diff
@@ -40,37 +127,67 @@ def get_match_result() -> str:
...
+        if not isinstance(result_json, dict):
+            raise Exception("Invalid match result format")
+
+        if "winner" not in result_json or "score" not in result_json:
+            raise Exception("Missing required fields in match result")
+
+        try:
+            winner = int(result_json["winner"])
+        except:
+            raise Exception("Invalid winner value")
```
[the remark this answers is not in the document]
Comment (linnbenton) [reply contracts/football_bets.py]: Screenshot_152

Good catch — updated to use specific exception types and proper exception chaining (from None) 👍

Screenshot_151
Also verified that all tests are passing locally.

File: contracts/football_bets.py [hunk 3]
```diff
@@ -40,37 +98,65 @@ def get_match_result() -> str:
...
+        except (ValueError, TypeError):
+            raise Exception("Invalid winner value")
+
+        score = result_json["score"]
+        if not isinstance(score, str):
+            raise Exception("Invalid score value")
+
+        return {
+            "winner": winner,
+            "score": score
```
[the remark this answers is not in the document]
Comment (linnbenton) [reply contracts/football_bets.py:172]: Screenshot_155

Added validation to restrict winner values to the supported enum, preventing invalid states 👍

[the remark this answers is not in the document]
Comment (linnbenton) [reply contracts/football_bets.py:172]: Thanks! Glad the fix aligns well 🙌

File: contracts/football_bets.py [hunk 4]
```diff
@@ -2,8 +2,57 @@
...
 
+    def allow_storage(cls):
+        return cls
+
+    class u256(int):
+        pass
+
+    class Address(str):
+        def as_hex(self):
+            return self
```
[the remark this answers is not in the document]
Comment (linnbenton) [reply contracts/football_bets.py:20]: Screenshot_159

Good catch — aligned Address.as_hex fallback with expected usage and ensured winner values are properly validated 👍

[the remark this answers is not in the document]
Comment (linnbenton) [reply contracts/football_bets.py:20]: Thanks for the review and feedback — really helpful! 🙌

**Comment by linnbenton** [comment]
Updated JSON parsing to use JSONDecodeError and verified via tests. All checks passing.

**Comment by JoaquinBN** [comment]
Thanks for the effort on validation and testing. Closing because this is stale and conflicts with `v2-dev`, embeds a fallback GenLayer runtime inside the contract, and uses the old test layout. Direct-test infrastructure and winner-range validation have since landed. Any remaining missing-bet/error handling should be proposed as a small current-line PR.
